system/Encryption/Encryption.php

```diff
@@ -0,0 +1,298 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @package    CodeIgniter
+ * @author     CodeIgniter Dev Team
+ * @copyright  2014-2017 British Columbia Institute of Technology (https://bcit.ca/)
+ * @license    https://opensource.org/licenses/MIT	MIT License
+ * @link       https://codeigniter.com
+ * @since      Version 4.0.0
+ * @filesource
+ */
+
+namespace CodeIgniter\Encryption;
+
+use Config\Encryption as EncryptionConfig;
+use CodeIgniter\Encryption\Exceptions\EncryptionException;
+use CodeIgniter\Config\Services;
+
+/**
+ * CodeIgniter Encryption Manager
+ *
+ * Provides two-way keyed encryption via PHP's Sodium and/or OpenSSL extensions.
+ * This class determines the driver, cipher, and mode to use, and then
+ * initializes the appropriate encryption handler.
+ */
+class Encryption
+{
+
+	/**
+	 * The encrypter we create
+	 *
+	 * @var string
+	 */
+	protected $encrypter;
+
+	/**
+	 * Our remembered configuration
+	 */
+	protected $config = null;
+
+	/**
+	 * Our default configuration
+	 */
+	protected $default = [
+		'driver' => 'OpenSSL', // The PHP extension we plan to use
+		'key'    => '', // no starting key material
+	];
+	protected $driver, $key;
+
+	/**
+	 * HMAC digest to use
+	 */
+	protected $digest = 'SHA512';
+
+	/**
+	 * Map of drivers to handler classes, in preference order
+	 *
+	 * @var array
+	 */
+	protected $drivers = [
+		'OpenSSL',
+	];
+
+	// --------------------------------------------------------------------
+
+	/**
+	 * Class constructor
+	 *
+	 * @param  mixed $params Configuration parameters
+	 * @return void
+	 *
+	 * @throws \CodeIgniter\Encryption\Exceptions\EncryptionException
+	 */
+	public function __construct($params = null)
+	{
+		$this->config = array_merge($this->default, (array) new \Config\Encryption());
+
+		if (is_string($params))
+		{
+			$params = ['driver' => $params];
+		}
+
+		$params = $this->properParams($params);
+
+		// Check for an unknown driver
+		if (isset($this->drivers[$params['driver']]))
+		{
+			throw EncryptionException::forDriverNotAvailable($params['driver']);
+		}
+
+		// determine what is installed
+		$this->handlers = [
```

Comment:
This way of checking doesn't support users creating their own handlers, does it? Is that something we should allow for this? Part of me says yes, because if they're on 7.3+ and the framework is still 7.2, they could write one to take advantage of new features.

Reply:
it wasn't the intent that users create their own drivers for use here, but that we might provide several alternatives